### PR TITLE
[SE] sync/async dispatchers

### PR DESCRIPTION
**Pull request by iceseer**
### Description of the Change
Async dispatcher implementation. Manager now contains IDispatcher reference and we can change it to sync one for tests.

### CODE REVIEW

File: irohad/subscription/dispatcher.hpp [hunk 1]
```diff
@@ -12,90 +12,27 @@
 namespace iroha::subscription {
 
   template <uint32_t kCount, uint32_t kPoolSize>
-  class Dispatcher final : utils::NoCopy, utils::NoMove {
-   public:
-    static constexpr uint32_t kHandlersCount = kCount;
-    static constexpr uint32_t kPoolThreadsCount = kPoolSize;
-    using Task = ThreadHandler::Task;
+  struct IDispatcher {
```
Comment (kuvaldini) [inline irohad/subscription/dispatcher.hpp:15]: I'd suggest to make constexpr constructor with these parameters. And keep IDispatcher not a template.

File: irohad/subscription/async_dispatcher_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        h.is_temporary = false;
+      }
+      for (auto &h : pool_) {
+        h.handler = std::make_shared<ThreadHandler>();
+        h.is_temporary = false;
+      }
+    }
+
+    void dispose() override {
+      for (auto &h : handlers_) h.handler->dispose();
```
Comment (kuvaldini) [inline irohad/subscription/async_dispatcher_impl.hpp:63]: Please format.

File: irohad/subscription/async_dispatcher_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      for (auto &h : pool_) h.handler->dispose();
```
Comment (kuvaldini) [inline irohad/subscription/async_dispatcher_impl.hpp:64]:
```suggestion
      for (auto &h : handlers_) 
        h.handler->dispose();
      for (auto &h : pool_)
        h.handler->dispose();
```

**Review by kuvaldini — APPROVED** [review]

File: irohad/subscription/dispatcher.hpp (hunk 1 above)
Comment (iceseer) [reply irohad/subscription/dispatcher.hpp:15]: IDispatcher is a template to make compile-time check of a executor thread in subscription calls.

File: irohad/subscription/async_dispatcher_impl.hpp (hunk 2 above)
Comment (iceseer) [reply irohad/subscription/async_dispatcher_impl.hpp:63]: it's formatted

File: test/module/irohad/subscription/subscription_test.cpp [hunk 4]
```diff
@@ -17,7 +18,8 @@ class SubscriptionTest : public ::testing::Test {
   template <uint32_t ThreadsCount>
   auto createSubscriptionManager() {
     using Manager = subscription::SubscriptionManager<ThreadsCount, 1u>;
-    return std::make_shared<Manager>();
+    return std::make_shared<Manager>(
+        std::make_shared<subscription::AsyncDispatcher<ThreadsCount, 1u>>());
```
[the remark this answers is not in the document]
Comment (iceseer) [reply test/module/irohad/subscription/subscription_test.cpp:22]: agree

File: irohad/subscription/async_dispatcher_impl.hpp (hunk 3 above)
Comment (iceseer) [reply irohad/subscription/async_dispatcher_impl.hpp:64]: it's formatted

File: irohad/subscription/dispatcher.hpp [hunk 5]
```diff
@@ -12,90 +12,27 @@
...
-      for (auto &h : handlers_) {
-        h.handler = std::make_shared<ThreadHandler>();
-        h.is_temporary = false;
-      }
-      for (auto &h : pool_) {
-        h.handler = std::make_shared<ThreadHandler>();
-        h.is_temporary = false;
-      }
-    }
+    static constexpr Tid kExecuteInPool = std::numeric_limits<Tid>::max();
```
Comment (lebdron) [inline irohad/subscription/dispatcher.hpp:19]: Please somehow make this accessible without class templates.

File: irohad/subscription/dispatcher.hpp [hunk 6]
```diff
@@ -12,90 +12,27 @@
...
+    static constexpr uint32_t kHandlersCount = kCount;
```
Comment (lebdron) [inline irohad/subscription/dispatcher.hpp:20]: IDispatcher should not have these parameters. They are not applicable with MockDispatcher, or single threaded dispatcher, for example.

**Review by lebdron — CHANGES_REQUESTED** [review]
Please move template parameters from the interface to AsyncDispatcher.

File: irohad/main/CMakeLists.txt [hunk 7]
```diff
@@ -35,7 +35,9 @@ target_link_libraries(pg_connection_init
     logger
     )
 
-add_library(subscription impl/subscription.cpp)
+add_library(subscription
+    impl/subscription.cpp
+    impl/dispatcher.cpp)
 target_link_libraries(subscription
     )
```
Comment (lebdron) [inline irohad/main/CMakeLists.txt:42]:
```suggestion
```

**Review by lebdron — APPROVED** [review]
Please make `test_subscription` target with sync dispatcher and `getSubscription()` symbol.

File: irohad/main/CMakeLists.txt [hunk 8]
```diff
@@ -35,8 +35,16 @@ target_link_libraries(pg_connection_init
     logger
     )
 
-add_library(subscription impl/subscription.cpp)
-target_link_libraries(subscription
+add_library(async_subscription
+    impl/subscription.cpp
+    impl/async_dispatcher.cpp)
+target_link_libraries(async_subscription
+    )
```
Comment (lebdron) [inline irohad/main/CMakeLists.txt:42]:
```suggestion
```

File: irohad/main/CMakeLists.txt [hunk 9]
```diff
@@ -35,8 +35,16 @@ target_link_libraries(pg_connection_init
...
+
+add_library(sync_subscription
+    impl/subscription.cpp
+    impl/sync_dispatcher.cpp)
+target_link_libraries(sync_subscription
     )
```
Comment (lebdron) [inline irohad/main/CMakeLists.txt:48]:
```suggestion
```
These make no effect

**Pull request merged**